### PR TITLE
[REVIEW] CSV Reader: Support ARROW_RANDOM_FILE input

### LINKED ISSUE #2493: [FEA] Add ARROW_RANDOM_ACCESS_FILE support to read_csv
Would like to have  ARROW_RANDOM_ACCESS_FILE support to read_csv the way that read_parquet does. This would allow BlazingSQL to use HDFS or S3 data sources (which which we use arrow::io::RandomAccessFile)  and only read as much of the file as necessary, or allow read_csv to read as it parses. Right now we are forced to always read the whole file prior to processing.

### PR DESCRIPTION
**Pull request by j-ieong**
Fixes #2493 and #2800.

The PR involves moving the C++ reader interface to using the same input design as the Parquet and ORC readers.

* Add support for ARROW_RANDOM_FILE to match Parquet and ORC readers
* Fixed bug with extra memory copies when using host buffer input
* Move input types to ctor in C++ CSV reader class
* Remove byte range and rows from being class members
* Move parsing options to ctor

### CODE REVIEW

File: python/cudf/cudf/_lib/csv.pyx [hunk 1]
```diff
@@ -207,10 +188,34 @@ cpdef read_csv(
...
+            new csv_reader(str(filepath_or_buffer).encode(), args)
+        )
+    elif buffer.shape[0] != 0:
+        reader = unique_ptr[csv_reader](
+            new csv_reader(<char *>&buffer[0], buffer.shape[0], args)
+        )
+    else:
+        reader = unique_ptr[csv_reader](
+            new csv_reader(<char *>NULL, 0, args)
+        )
```
Comment (kkraus14) [inline python/cudf/cudf/_lib/csv.pyx]: You're no longer releasing the GIL here which you should be, the calls to the `csv_reader` should be in a `with nogil` context

File: python/cudf/cudf/_lib/csv.pyx [hunk 2]
```diff
@@ -207,10 +188,34 @@ cpdef read_csv(
...
-        reader = unique_ptr[cpp_csv.reader](new cpp_csv.reader(args))
+    # Create reader from source
+    cdef unique_ptr[csv_reader] reader
+    cdef const unsigned char[::1] buffer = None
+    if isinstance(filepath_or_buffer, BytesIO):
+        buffer = filepath_or_buffer.getbuffer()
+    elif isinstance(filepath_or_buffer, StringIO):
+        buffer = filepath_or_buffer.read().encode()
+    elif isinstance(filepath_or_buffer, bytes):
+        buffer = filepath_or_buffer
```
Comment (kkraus14) [inline python/cudf/cudf/_lib/csv.pyx]: Should this be put into a utility function and shared across the JSON reader and any other readers?

File: python/cudf/cudf/_lib/csv.pyx (hunk 1 above)
Comment (j-ieong) [reply python/cudf/cudf/_lib/csv.pyx]: Done, along with the read() calls as well.

File: python/cudf/cudf/_lib/csv.pyx (hunk 2 above)
Comment (j-ieong) [reply python/cudf/cudf/_lib/csv.pyx]: Yup, good idea

**Comment by mjsamoht** [comment]
@harrism the arrow file source test has been added.  Let us know if you're okay with that.

**Pull request merged**